### PR TITLE
Make CollectSequencingArtifactMetrics faster

### PR DESCRIPTION
**Pull request by meganshand**
### Description

This makes CollectSequencingArtifactMetrics faster by reducing the number of HashMap.get() functions we call at every base of the traversal. Previously we counted each base of each read into several HashMaps for different kinds of contexts ("full", "leading", "trailing", and "zero"). However, "leading", "trailing" and "zero" can all be determined from the "full" counts at the end of traversing the bam. This reduces the runtime on a 30X whole genome when `CONTEXT_SIZE=1` (the default):

image

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 1]
```diff
@@ -41,7 +42,47 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
+        for(Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()){
+            final String context = fullContext.getKey();
+            final char centralBase = context.charAt(contextSize);
+            final String leadingKey = context.substring(0, contextSize) + centralBase + padding;
+            final String trailingKey = padding + centralBase + context.substring(contextSize + 1, context.length());
+
+            AlignmentAccumulator[] trailingAlignmentAccumulators = this.artifactMap.get(trailingKey);
+            AlignmentAccumulator[] leadingAlignmentAccumulators = this.artifactMap.get(leadingKey);
+
+            for(int i=0; i < fullContext.getValue().length; i++){
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: extract value into a final AlignmentAccumulator[] variable at the top of the loop.

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 2]
```diff
@@ -136,5 +177,12 @@ private void countRecord(final SAMRecord rec) {
                 else this.R1_POS++;
             }
         }
+
+        private void includeCounts(final AlignmentAccumulator other){
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: This type of function is usually called `merge`.

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 3]
```diff
@@ -41,7 +42,47 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
+    public void fillZeroRecords(final ContextAccumulator fullContextAccumulator, int contextSize) {
+        final String padding = StringUtil.repeatCharNTimes('N', contextSize);
+        for(Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()){
+            final String context = fullContext.getKey();
+            final char centralBase = context.charAt(contextSize);
+            final String key = padding + centralBase + padding;
+
+            AlignmentAccumulator[] alignmentAccumulators = this.artifactMap.get(key);
+
+            for(int i=0; i < fullContext.getValue().length; i++){
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: here too (extract)

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 4]
```diff
@@ -41,7 +42,47 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
     }
 
     public void countRecord(final String refContext, final char calledBase, final SAMRecord rec) {
-        artifactMap.get(refContext)[Transition.baseIndexMap[calledBase]].countRecord(rec);
+        AlignmentAccumulator[] accumulators = artifactMap.get(refContext);
+        if(accumulators != null) {accumulators[Transition.baseIndexMap[calledBase]].countRecord(rec);}
```
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: oops, forgot to reformat

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 5]
```diff
@@ -41,7 +42,47 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
+    }
+
+    /**
+     * Fills a halfContextAccumualtor by summing over the appropriate counts from a fullContextAccumulator.
```
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: fixed

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 6]
```diff
@@ -41,7 +42,47 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
+     */
+    public void fillHalfRecords(final ContextAccumulator fullContextAccumulator, int contextSize){
```
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 1 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 3 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 2 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: Thank you, that's the word I was looking for!

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 7]
```diff
@@ -43,44 +43,50 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
-            AlignmentAccumulator[] leadingAlignmentAccumulators = this.artifactMap.get(leadingKey);
+            final AlignmentAccumulator[] trailingAlignmentAccumulators = this.artifactMap.get(trailingKey);
+            final AlignmentAccumulator[] leadingAlignmentAccumulators = this.artifactMap.get(leadingKey);
+            final AlignmentAccumulator[] fullAlignmentAccumulators = fullContext.getValue();
 
-            for(int i=0; i < fullContext.getValue().length; i++){
-                trailingAlignmentAccumulators[i].includeCounts(fullContext.getValue()[i]);
-                leadingAlignmentAccumulators[i].includeCounts(fullContext.getValue()[i]);
+            for (int i=0; i < fullAlignmentAccumulators.length; i++) {
+                trailingAlignmentAccumulators[i].mergeCounts(fullContext.getValue()[i]);
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: fullContext.getValue() still here. should be fullAlignmentAccumulators

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 8]
```diff
@@ -178,7 +184,7 @@ private void countRecord(final SAMRecord rec) {
             }
         }
 
-        private void includeCounts(final AlignmentAccumulator other){
+        private void mergeCounts(final AlignmentAccumulator other) {
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: Given that this class is nothing but counts, I'd call this method `merge`, not `mergeCounts`

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 9]
```diff
@@ -43,44 +43,50 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
             final String key = padding + centralBase + padding;
 
-            AlignmentAccumulator[] alignmentAccumulators = this.artifactMap.get(key);
+            final AlignmentAccumulator[] zeroAlignmentAccumulators = this.artifactMap.get(key);
+            final AlignmentAccumulator[] fullAlignmentAccumulators = fullContext.getValue();
 
-            for(int i=0; i < fullContext.getValue().length; i++){
-                alignmentAccumulators[i].includeCounts(fullContext.getValue()[i]);
+            for (int i=0; i < fullAlignmentAccumulators.length; i++) {
+                zeroAlignmentAccumulators[i].mergeCounts(fullContext.getValue()[i]);
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: here too.

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 10]
```diff
@@ -43,44 +43,50 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
      */
-    public void fillZeroRecords(final ContextAccumulator fullContextAccumulator, int contextSize) {
+    public void fillZeroRecords(final ContextAccumulator fullContextAccumulator, final int contextSize) {
         final String padding = StringUtil.repeatCharNTimes('N', contextSize);
-        for(Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()){
+
+        for (Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()) {
             final String context = fullContext.getKey();
             final char centralBase = context.charAt(contextSize);
             final String key = padding + centralBase + padding;
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: rename key->zeroContextKey or similar

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java [hunk 11]
```diff
@@ -43,44 +43,50 @@ public ContextAccumulator(final Set<String> contexts, final boolean expectedTand
...
-     * Fills a zeroContextAccumualtor by summing over the appropriate counts from a fullContextAccumulator.
+     * Fills a zeroContextAccumulator by summing over the appropriate counts from a fullContextAccumulator.
      */
-    public void fillZeroRecords(final ContextAccumulator fullContextAccumulator, int contextSize) {
+    public void fillZeroRecords(final ContextAccumulator fullContextAccumulator, final int contextSize) {
         final String padding = StringUtil.repeatCharNTimes('N', contextSize);
-        for(Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()){
+
+        for (Map.Entry<String,AlignmentAccumulator[]> fullContext : fullContextAccumulator.artifactMap.entrySet()) {
             final String context = fullContext.getKey();
```
Comment (yfarjoun) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: rename context -> fullContextKey or similar

**Comment by yfarjoun** [comment]
I am not worried about the error metrics.

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 7 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: Oh I see it now. Fixed.

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 11 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 10 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 9 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

File: src/main/java/picard/analysis/artifacts/ContextAccumulator.java (hunk 8 above)
Comment (meganshand) [inline src/main/java/picard/analysis/artifacts/ContextAccumulator.java]: done

**Pull request merged**